test/e2e/framework/pod/wait.go

```diff
@@ -65,31 +65,73 @@ const (
 
 type podCondition func(pod *v1.Pod) (bool, error)
 
-// errorBadPodsStates create error message of basic info of bad pods for debugging.
-func errorBadPodsStates(badPods []v1.Pod, desiredPods int, ns, desiredState string, timeout time.Duration, err error) string {
-	errStr := fmt.Sprintf("%d / %d pods in namespace %q are NOT in %s state in %v\n", len(badPods), desiredPods, ns, desiredState, timeout)
-	if err != nil {
-		errStr += fmt.Sprintf("Last error: %s\n", err)
+type timeoutError struct {
+	msg string
+}
+
+func (e *timeoutError) Error() string {
+	return e.msg
+}
+
+func TimeoutError(format string, args ...interface{}) *timeoutError {
+	return &timeoutError{
+		msg: fmt.Sprintf(format, args...),
+	}
+}
+
+// MaybeTimeoutError returns a TimeoutError if err is a timeout. Otherwise, wrap err.
+// taskFormat and taskArgs should be the task being performed when the error occurred,
+// e.g. "waiting for pod to be running".
+func MaybeTimeoutError(err error, taskFormat string, taskArgs ...interface{}) error {
+	if IsTimeout(err) {
+		return TimeoutError("timout "+taskFormat, taskArgs...)
```

Comment:
IMO nested polling is an anti-pattern, and should be avoided. That said, I could imagine a non-test function that get's called in a polling wait function, and returns `wait.ErrWaitTimeout`. However, given the following nested polling, there isn't any way to tell which loop timed out (going back to the anti-pattern...):

```go
outerErr := wait.Poll(poll, timeout, func() (bool, error) {
  err := wait.Poll(poll, otherTimeout, func() (bool, error) {  // might be obscured inside another function
    // ...
  })
  if err != nil {
    return false, err
  }
  // ...
})
```
In this example, if `outerErr == wait.ErrWaitTimeout` there's no way to tell which `wait.Poll` timed out, unless the inner err gets wrapped. Given that, maybe the solution is to just not unwrap errors in `IsTimeout`, and only look at the outer error?

Reply:
`timout` was definitely not intentional :laughing: I'll go with `timed out while ...` instead

Reply:
Darn, I hadn't even noticed the typo :grin: 

> Given that, maybe the solution is to just not unwrap errors in IsTimeout, and only look at the outer error?

I think I prefer such a simpler solution because the logic and effect in `MaybeTimeoutError` will be much easier to understand.